### PR TITLE
[WIP][Finetune] Report training metrics to Tensorboard

### PR DESCRIPTION
**Pull request by carsonwang**
When you start finetuning, you will see a message like below:
To visualize your results with TensorBoard, run: `tensorboard --logdir /xxx/ray_results/TorchTrainer_2024-01-08_16-51-19`

This PR reports metrics train loss and perplexity to Ray that writes to TensorBoard.
Also introduced parameter logging_steps to control the logging frequency.

### CODE REVIEW

File: common/trainer/default_trainer.py [hunk 1]
```diff
@@ -130,7 +131,7 @@ def prepare(self, model, tokenizer, dataset, optimizer, accelerator):
     def train(self):
         num_train_epochs = self.config.get("num_train_epochs", 1)
         checkpoint = self.config.get("checkpoint")
-        log_step = self.config.get("log_step", 1)
+        logging_steps = self.config.get("logging_steps")
```
Comment (KepingYan) [inline common/trainer/default_trainer.py:134]: Please add default value `1` because of ui.

File: common/trainer/default_trainer.py [hunk 2]
```diff
@@ -147,12 +148,19 @@ def train(self):
...
-                    if step % log_step == 0:
-                        logger.info(f"train epoch:[{idx}/{num_train_epochs}]\tstep:[{step}/{total_steps}]\tloss:{loss:.6f}\tppl:{math.exp(loss):.6f}\ttime:{time.time()-start:.6f}")
-                        report({"train_epoch": idx, "total_epochs": num_train_epochs, "train_step": step, "total_steps": min(max_train_step, total_steps) if max_train_step else total_steps})
+
+                    # Checks if the accelerator has performed an optimization step behind the scenes
+                    if self.accelerator.sync_gradients:
+                        self.completed_steps += 1
+
+                    if self.completed_steps % logging_steps == 0:
+                        perplexity = math.exp(loss)
```
Comment (KepingYan) [inline common/trainer/default_trainer.py:157]: Would it be better to use `loss.item()` here?

File: common/trainer/default_trainer.py (hunk 1 above)
Comment (carsonwang) [reply common/trainer/default_trainer.py:134]: OK. I've set the default to 10 in the configuration file. Fine for UI?

File: common/trainer/default_trainer.py (hunk 2 above)
Comment (carsonwang) [reply common/trainer/default_trainer.py:157]: Yes, thanks. I will update.

**Comment by carsonwang** [comment]
I will also update the logging format and data.

File: common/trainer/default_trainer.py (hunk 1 above)
Comment (KepingYan) [reply common/trainer/default_trainer.py:134]: Got it. It‘s because now start_ui.py does not use finetune.yaml, so the 'logging_steps' will be none. Or I modify start_ui.py later to get the configuration from finetune.yaml.

File: common/trainer/default_trainer.py [hunk 3]
```diff
@@ -147,12 +148,19 @@ def train(self):
                     if self.lr_scheduler is not None:
                         self.lr_scheduler.step()
                     self.optimizer.zero_grad()
-                    if step % log_step == 0:
-                        logger.info(f"train epoch:[{idx}/{num_train_epochs}]\tstep:[{step}/{total_steps}]\tloss:{loss:.6f}\tppl:{math.exp(loss):.6f}\ttime:{time.time()-start:.6f}")
-                        report({"train_epoch": idx, "total_epochs": num_train_epochs, "train_step": step, "total_steps": min(max_train_step, total_steps) if max_train_step else total_steps})
```
Comment (KepingYan) [inline common/trainer/default_trainer.py:152]: This line is for updating the progress bar on web ui, the key of this dict can't be modified because of https://github.com/intel/llm-on-ray/blob/f26343d14320f0be889f46c3d1703a743de82f57/ui/start_ui.py#L82.

File: common/trainer/default_trainer.py (hunk 1 above)
Comment (carsonwang) [reply common/trainer/default_trainer.py:134]: I understood it. Sorry I didn't say clearly. I mean I will update this to `logging_steps = self.config.get("logging_steps", 10)` here using 10 instead of 1 as default.

File: common/trainer/default_trainer.py (hunk 3 above)
Comment (carsonwang) [reply common/trainer/default_trainer.py:152]: Thanks for the info. Let me double check how to update this. All the keys and values will also be shown in Tensorboard.

File: common/trainer/default_trainer.py (hunk 1 above)
Comment (KepingYan) [reply common/trainer/default_trainer.py:134]: OK, but progress bar needs the values of `report({})` to be updated every step, otherwise the status of progress bar will only be updated after 10 steps.

Comment (carsonwang) [reply common/trainer/default_trainer.py:134]: I see. Ok, will set it to 1 here.

**Comment by carsonwang** [comment]
I'll close this one. @harborn will continue to work on this and address the comments.
